src/ratelimiter_map.rs

```diff
@@ -0,0 +1,135 @@
+
+    loop {
+        interval.tick().await;
+        let right_now = Instant::now();
+
+        map.retain(|_, (_, last_used)| *last_used + client_decay_timeout > right_now);
+
+        debug!("Done reaping timed out HTTP ratelimiters");
+    }
+}
+
+impl RatelimiterMap {
+    pub fn new(mut default_token: String) -> Self {
+        let is_bot = default_token.starts_with("Bot ");
+        let is_bearer = default_token.starts_with("Bearer ");
+
+        // Make sure it is either a bot or bearer token, and assume it's a bot
+        // token if no prefix is given
+        if !is_bot && !is_bearer {
+            default_token.insert_str(0, "Bot ");
+        }
+
+        let max_size = var("CLIENT_CACHE_MAX_SIZE").map_or(None, |size| {
+            if let Ok(size) = size.parse() {
+                Some(size)
+            } else {
+                warn!("Unable to parse CLIENT_CACHE_MAX_SIZE, proceeding with defaults");
+                None
+            }
+        });
+
+        let inner = Arc::new(DashMap::new());
+        let default = Ratelimiter::new();
+
+        tokio::spawn(reap_old_ratelimiters(inner.clone()));
+
+        Self {
+            default,
+            default_token,
+            max_size,
+            inner,
+        }
+    }
+
+    fn lru(&self) -> Option<RefMulti<String, (Ratelimiter, Instant)>> {
+        self.inner.iter().next().map(|first_entry| {
+            self.inner.iter().fold(
+                first_entry,
+                |old, next| {
+                    if old.1 > next.1 {
+                        next
+                    } else {
+                        old
+                    }
+                },
+            )
+        })
+    }
+
+    pub fn get_or_insert(&self, token: Option<&str>) -> (Ratelimiter, String) {
+        if let Some(token) = token {
+            if token == self.default_token {
+                (self.default.clone(), self.default_token.clone())
+            } else {
+                let access_time = Instant::now();
+
+                if let Some(mut entry) = self.inner.get_mut(token) {
+                    entry.1 = access_time;
+                    (entry.0.clone(), token.to_string())
+                } else {
+                    if self
+                        .max_size
+                        .filter(|max_size| self.inner.len() >= *max_size && max_size > &0)
+                        .is_some()
+                    {
+                        let key = self
+                            .lru()
+                            .expect("Length of inner map is guaranteed to be greater than 0")
+                            .key()
+                            .clone();
+
+                        self.inner.remove(&key);
+                        debug!("Removed oldest entry from HTTP ratelimiter cache");
+                    }
+
+                    let ratelimiter = Ratelimiter::new();
+
+                    if self.max_size.filter(|max| max != &0).is_some() {
```

Comment:
Same with filter here

Reply:
Is there a purpose to a max size of 0? Doesn't that essentially make the proxy forget a token's ratelimiter state once a request handler has finished execution, thus removing the point of the proxy?